tests/PHPStan/Generics/AssertTypeRule.php

```diff
@@ -0,0 +1,59 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Generics;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Name;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\Rule;
+use PHPStan\Type\TypeUtils;
+use PHPStan\Type\VerbosityLevel;
+
+class AssertTypeRule implements Rule
+{
+
+	private const ASSERT_TYPE_FUNCTION = 'PHPStan\\Generics\\assertType';
+
+	public function getNodeType(): string
+	{
+		return FuncCall::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		assert($node instanceof FuncCall);
+
+		$nameNode = $node->name;
+
+		if (!$nameNode instanceof Name) {
+			return [];
+		}
+
+		if ((string) $nameNode !== self::ASSERT_TYPE_FUNCTION) {
+			return [];
+		}
+
+		if (count($node->args) !== 2) {
+			return [];
+		}
+
+		$typeType = $scope->getType($node->args[0]->value);
+		$valueType = $scope->getType($node->args[1]->value);
+
+		$constants = TypeUtils::getConstantStrings($typeType);
+		foreach ($constants as $constant) {
+			$typeString = $valueType->describe(VerbosityLevel::precise());
+			if ($constant->getValue() !== $typeString) {
+				return [sprintf(
+					'Expected type %s, got %s',
```

Comment:
I don't think this rule is a good fit inside this PR. There's a similar proposal here (https://github.com/phpstan/phpstan/pull/2102) but I unfortunately haven't had time to look at it. Everything around asserting types should be done in NodeScopeResolverTest until we have a better complex approach in master. For example look at dataBinary/testBinary methods in NodeScopeResolverTest to get an idea how it works. Please add any type assertions in new dataGenerics/testGenerics pair there, thank you.

Reply:
You can either have a single assertion point at `die;` or you can specify your own assertion points like it's in `testForeachLoopVariables`.